### ISSUE TITLE
OS supported

### ISSUE THREAD

**Issue by SirWindfield** [jaemk/self_update#21]
Which OS does this library support? IIRC Windows doesn't allow replacing binaries at runtime because they get locked when a program uses them. 

My wild guess would be that `self_update` doesn't support Windows.

**Comment by jaemk** [comment]
I don't have a windows device to test on so I'm not positive how it behaves. The appveyor test script could be updated to better test this

**Comment by armoha** [comment]
Example code has `ZipError: Specified file not found in archive` on Windows 10 x64.
```
...
Checking target-arch... x86_64-pc-windows-msvc
Checking current version... v0.5.1
Checking latest released version... v9.9.10
New release found! v0.5.1 --> v9.9.10
New release is *NOT* compatible

github release status:
  * Current exe: "C:\\Users\\artan\\CLionProjects\\self_update\\target\\debug\\examples\\github.exe"
  * New exe release: "self_update-v9.9.10-x86_64-pc-windows-msvc.zip"
  * New exe download url: "https://github.com/jaemk/self_update/releases/download/v9.9.10/self_update-v9.9.10-x86_64-pc-windows-msvc.zip"

The new release will be downloaded/extracted and the existing binary will be replaced.
Do you want to continue? [Y/n] Y
Downloading...
1.09 MB / 1.09 MB [=====================================] 100.00 % 481.01 KB/s   ... Done
Extracting archive... [ERROR] ZipError: Specified file not found in archive
error: process didn't exit successfully: `target\debug\examples\github.exe` (exit code: 1)

Process finished with exit code 1
```
Changing `.bin_name("github")` to `.bin_name("github.exe")` works well, successfully update executable.
```powershell
...
PS C:\Users\artan\CLionProjects\self_update\target\debug\examples> .\github.exe
Checking target-arch... x86_64-pc-windows-msvc
Checking current version... v9.9.10
Checking latest released version... v9.9.10
Update status: `9.9.10`!
```

**Comment by r-darwish** [comment]
I tested this crate on Windows, Linux, macOS and FreeBSD. It works on all of them.

**Comment by SirWindfield** [comment]
> Changing `.bin_name("github")` to `.bin_name("github.exe")` works well, successfully update executable.

Wouldn't this still be a problem on Windows? :thinking:

**Comment by canselcik** [comment]
That would be my expectation as well. After all Windows will not allow deleting or opening to edit a PE if it is the image for a currently running process.

**Comment by r-darwish** [comment]
On Windows we move the old binary to the TEMP folder, where it will be cleaned by Windows’ disk cleanup utility

**Comment by jaemk** [comment]
Here's the replacement logic that does the swap using a temp file https://github.com/jaemk/self_update/blob/ef85a3451f27965f9d1f60eb080cbb348efd0149/src/lib.rs#L427

The `.exe` extension issue will hopefully be resolved by #30 . @SirWindfield I can finish that up for you if you're too busy, but I don't want to pull it out from under you if you were planning on continuing with it!

**Comment by SirWindfield** [comment]
It's ok. I don't have time right now due to exams. Feel free to update the PR. I'll just tackle down something else ;)
